sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUuidColumnsOfProjects.java

```diff
@@ -0,0 +1,167 @@
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.db.Database;
+import org.sonar.db.version.BaseDataChange;
+import org.sonar.db.version.MassUpdate;
+import org.sonar.db.version.Select;
+import org.sonar.db.version.SqlStatement;
+
+public class PopulateUuidColumnsOfProjects extends BaseDataChange {
+  private static final Logger LOG = Loggers.get(PopulateUuidColumnsOfProjects.class);
+
+  public PopulateUuidColumnsOfProjects(Database db) {
+    super(db);
+  }
+
+  @Override
+  public void execute(Context context) throws SQLException {
+
+    Map<Long, String> componentUuidById = buildComponentUuidMap(context);
+    if (componentUuidById.isEmpty()) {
+      return;
+    }
+
+    populateRootUuidColumnForRoots(context);
+    populateRootUuidColumnForSubnodes(context, componentUuidById);
+    populateCopyComponentUuidColumn(context, componentUuidById);
+    populatePersonUuidColumn(context, componentUuidById);
+  }
+
+  private static Map<Long, String> buildComponentUuidMap(Context context) throws SQLException {
+    Map<Long, String> componentUuidById = new HashMap<>();
+    // rootId for root nodes (ie. column root_id is null)
+    context.prepareSelect("select distinct p1.id, p1.uuid from projects p1" +
+        " where p1.root_id is null and p1.root_uuid is null")
+        .scroll(row -> componentUuidById.put(row.getLong(1), row.getString(2)));
+    // rootId for other nodes (ie. column root_id is not null)
+    context.prepareSelect("select distinct p1.id, p1.uuid from projects p1" +
+      " join projects p2 on p1.id = p2.root_id" +
+      " where p2.root_uuid is null")
+      .scroll(row -> componentUuidById.put(row.getLong(1), row.getString(2)));
+    // copyResourceId
+    context.prepareSelect("select distinct p1.id, p1.uuid from projects p1" +
+      " join projects p2 on p1.id = p2.copy_resource_id" +
+      " where p2.copy_resource_id is not null and p2.copy_component_uuid is null")
+      .scroll(row -> componentUuidById.put(row.getLong(1), row.getString(2)));
+    // person_id
+    context.prepareSelect("select distinct p1.id, p1.uuid from projects p1" +
+      " join projects p2 on p1.id = p2.person_id" +
+      " where p2.person_id is not null and p2.developer_uuid is null")
+      .scroll(row -> componentUuidById.put(row.getLong(1), row.getString(2)));
+    return componentUuidById;
+  }
+
+  private void populateRootUuidColumnForRoots(Context context) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("SELECT distinct p.id from projects p where p.root_id is null and p.root_uuid is null");
+    massUpdate.update("UPDATE projects SET root_uuid=uuid WHERE id=? and root_id is null and root_uuid is null");
+    massUpdate.rowPluralName("root uuid of root components");
+    massUpdate.execute(this::handleRootIdUpdateForRootNodes);
+  }
+
+  private boolean handleRootIdUpdateForRootNodes(Select.Row row, SqlStatement update) throws SQLException {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make "handleRootIdUpdateForRootNodes" a "static" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Private method 'handleRootIdUpdateForRootNodes' is never used. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)